### DISCUSSION TITLE
Python: can't add more than 1 scope?

### DISCUSSION THREAD

**Discussion by geolectric** [etsy/open-api#760] [Q&A]
I'm using Python, and I started out just using the 'listings_w' scope, and everything worked fine.

Then I tried adding: listings_r, transactions_r, and transactions_w:

```
resp = requests.get(
    url="https://www.etsy.com/oauth/connect",
    params={
        "response_type": "code",
        "redirect_uri": "https://localhost:8080/queue",
        "scope": "listings_r%20listings_w%20transactions_r%20transactions_w",
        "client_id": etsy_api_key,
        "state": etsy_state,
        "code_challenge": code_challenge,
        "code_challenge_method": "S256",
    }
)

return redirect(resp.url)
```

But it just keeps kicking me back to the same login screen, it never redirects me to the next page.

However, if I use any of the above scopes by itself, then it works and shows me the Grant Access page.

What am I doing wrong here?

**Comment by etsyachristensen** [comment]
Is this on your own shop just in testing? If so, just for the sake of narrowing down the possibilities, can you go to https://www.etsy.com/your/apps and see if the app is listed under "Third Party Apps to which you've granted access". If it is, try revoking the grant and then try your full string of scopes again. I'm curious if there's something up with how we check for previous grant requests. Please reply back with whether or not it worked.

**Reply by geolectric** [reply]
@etsyachristensen I had actually already tried that. I found that screen and there were multiple approvals for the same app, probably from when I was testing. I ended up Revoking them all. And I was still having the above issue. Currently, there's just 1 approved from when I was testing again with just 1 scope, which works. But even after revoking everything I still couldn't add more than 1.

**Reply by etsyachristensen** [reply]
Ok thank you for the confirmation. We'll get this on our list of issues to investigate. Sorry for the trouble!

**Comment by geolectric** [comment]
Any updates on this or ideas to try? Does anyone else have this issue? My account is in developer mode... Could that matter?

**Reply by baahar** [reply]
I would suggest looking at the generated URL that contains the parameters. It could be an encoding issue regarding the %20 character.

**Comment by 2003crimson** [comment]
I have the same problem

Here is my requst url 

 redirect_uri:https://www.etsy.com/oauth/connect?response_type=code&redirect_uri=https://www.baidu.com&scope=address_r%20address_w%20billing_r%20cart_r%20cart_w%20email_r%20favorites_r%20favorites_w%20feedback_r%20listings_d%20listings_r%20listings_w%20profile_r%20profile_w%20recommend_r%20recommend_w%20shops_r%20shops_w%20transactions_r%20transactions_w&client_id=keySting&state=superstate&code_challenge=DSWlW2Abh-cf8CeLL8-g3hQ2WQyYdKyiu83u_s7nRhI&code_challenge_method=S256

But i got www.baidu.com?error=invalid_scope&error_description=scope+is+invalid&state=superstate

And when i used just only one scope without '%20' or a blank space,  i succeeded

Please tell me if there is sth wrong with the url

**Reply by 2003crimson** [reply]
@etsyachristensen Could you please take a look and help me figure out what the issue is?

**Comment by Wilderbill** [comment]
I am having a similar if not the same problem:
This works:        scope=listings_r%20email_r
This does not:   scope=listings_r%20listings_w
I want to add the listings_w scope so that I can modify listings with my code.

**Reply by geolectric** [reply]
I haven't looked at this for years as I'm not working on this project anymore, however, here's what I last had that worked. You could probably create a cleaner solution but notice the `replace` in the redirect url.

```
etsy_api_key = get_etsy_api_key()
etsy_state = get_etsy_state()

code_verifier, code_challenge = pkce.generate_pkce_pair()

set_etsy_code_verifier_and_challenge(code_verifier, code_challenge)

resp = requests.get(
    url="https://www.etsy.com/oauth/connect",
    params={
        "response_type": "code",
        "redirect_uri": "https://localhost:5000/queue",
        "scope": "address_r%20address_w%20billing_r%20cart_r%20cart_w%20email_r%20favorites_r%20favorites_w%20feedback_r%20listings_d%20listings_r%20listings_w%20profile_r%20profile_w%20recommend_r%20recommend_w%20shops_r%20shops_w%20transactions_r%20transactions_w",
        "client_id": etsy_api_key,
        "state": etsy_state,
        "code_challenge": code_challenge,
        "code_challenge_method": "S256",
    }
)

return redirect(resp.url.replace("%252520", "%20"))
```

You might need to do something different, but I would look very closely at the URL returned by Etsy, which for me needed modified.

**Comment by Wilderbill** [comment]
Okay, I wanted to be able to read and write my Etsy listings.
Copied code from:  https://github.com/slider1234/ETSY-API-Auto-Authenticator
That had:   scope = "listings_r email_r" and that worked for reading.
First guess for writing was:  scope = "listings_r listings_w" ... raised an error.
Second guess was:  scope = "listings_r listings_w email_r" ... raised an error.
Third guess was:  scope = "listings_w email_r" and that worked.
So, apparent email_r is mandatory, and listings_w implies listings _r, but you can't specify both.

**Reply by Wilderbill** [reply]
Let me retract that previous posting.
I believe I now have it working, again.
The problem was multiple overlapping instances as revealed by "Apps You Use".
(  See the website: https://www.etsy.com/your/apps?ref=si_apps  )
I had a half dozen active instances of my app as a result of my flailing attempts to get it working.
Apparently, there can be multiple instances of your app with overlapping, conflicting scopes.
The solution was to delete all of my instances and start again.
Set the scopes you want and run: https://github.com/slider1234/ETSY-API-Auto-Authenticator again.
(  This does not require a repeat of the 3-week application process.  )
Bill

**Comment by ramibch** [comment]
Use the PyPI package etsyv3 (Check out the class `etsyv3.util.auth.AuthHelper`)

**Reply by ramibch** [reply]
here is a small integration with a django project: https://github.com/ramibch/one/blob/main/one/etsy/views.py
